src/GovUk.Education.ExploreEducationStatistics.Admin/Services/UpdateChartFilesService.cs

```diff
@@ -0,0 +1,168 @@
+            {
+                var oldName = fileReference.Filename;
+                var oldBlobPath = AdminReleasePathWithFilename(fileReference.ReleaseId, oldName);
+
+                // Files should exists first time this is run in storage
+                if (blobContainer.GetBlockBlobReference(oldBlobPath).Exists())
+                {
+                    if (await CopyBlob(fileReference, blobContainer, oldBlobPath, fileReference.Id))
+                    {
+                        // Change ref to any datablocks with charts where used
+
+                        await UpdateDataBlock(fileReference, fileReference.Id);
+
+                        await _contentDbContext.SaveChangesAsync();
+
+                        _logger.LogInformation(
+                            $"Update blob: {fileReference.ReleaseId}:{fileReference.Filename} successfully");
+                    }
+                }
+            }
+
+            return true;
+        }
+
+        private async Task<bool> CopyBlob(ReleaseFileReference fileReference, CloudBlobContainer blobContainer, string oldBlobPath, Guid newId)
+        {
+            var newBlobPath = AdminReleasePathWithFilename(fileReference.ReleaseId, newId.ToString());
+
+            var sourceBlob = blobContainer.GetBlockBlobReference(oldBlobPath);
+            var targetBlob = blobContainer.GetBlockBlobReference(newBlobPath);
+                        
+            await targetBlob.StartCopyAsync(sourceBlob);
+            targetBlob.FetchAttributes();
+            while (targetBlob.CopyState.Status == CopyStatus.Pending) {
+                await Task.Delay(1000);
+                targetBlob.FetchAttributes();
+            }
+
+            if (targetBlob.CopyState.Status != CopyStatus.Success) {
+                _logger.LogError($"Failed to copy blob: {fileReference.ReleaseId}:{fileReference.Filename}");
+                return false;
+            }
+
+            await sourceBlob.DeleteAsync();
+            
+            _logger.LogInformation($"Copied blob: {fileReference.ReleaseId}:{fileReference.Filename} successfully");
+
+            return true;
+        }
+        
+        private async Task UpdateDataBlock(ReleaseFileReference releaseFileReference, Guid newChartId)
+        {
+            var releaseIds = await _contentDbContext.ReleaseFiles
+                .Where(rf => rf.ReleaseFileReferenceId == releaseFileReference.Id)
+                .Select(rf => rf.ReleaseId)
+                .Distinct()
+                .ToListAsync();
+            
+            var blocks = GetDataBlocks(releaseIds);
+
+            foreach (var block in blocks)
+            {
+                var infoGraphicChart = block.Charts
+                    .OfType<InfographicChart>()
+                    .FirstOrDefault();
+
+                if (infoGraphicChart != null && infoGraphicChart.FileId == releaseFileReference.Filename)
+                {
+                    block.Charts.Remove(infoGraphicChart);
```

Comment:
Do we need the remove/add? We've already got the reference to `infoGraphicChart` so couldn't we just mutate it directly?

Reply:
no needs to be like that :)